std/algorithm/searching.d

```diff
@@ -1773,44 +1773,85 @@ if (isRandomAccessRange!R1 && hasLength!R1 && hasSlicing!R1 && isBidirectionalRa
         && is(typeof(binaryFun!pred(haystack.front, needle.front)) : bool))
 {
     if (needle.empty) return haystack;
-    const needleLength = walkLength(needle.save);
-    if (needleLength > haystack.length)
+    static if (hasLength!R2)
     {
-        return haystack[haystack.length .. haystack.length];
+        immutable needleLength = needle.length;
     }
-    // @@@BUG@@@
-    // auto needleBack = moveBack(needle);
-    // Stage 1: find the step
-    size_t step = 1;
-    auto needleBack = needle.back;
-    needle.popBack();
-    for (auto i = needle.save; !i.empty && i.back != needleBack;
-         i.popBack(), ++step)
+    else
     {
+        immutable needleLength = walkLength(needle.save);
```

Comment:
@JackStouffer imagine a forward range that has a relatively cheap `.length`, but an expensive `.save`. This is the case I think @qznc has optimized for.

IMO, I don't think the optimization is worth it, since most cases of save are trivially just `return this;`. But the logic of having it is ok.